shimmingtoolbox/dicom_to_nifti.py

```diff
@@ -0,0 +1,134 @@
+import os
+import numpy as np
+from .read_nii import read_nii
+
+
+def dicom_to_nifti(unsortedDicomDir, niftiPath):
+    """
+    TODO: indicate the purpose of this function
+    :param unsortedDicomDir:
+    :param niftiPath:
+    :return:
+    """
+
+    #os.mkdir(niftiPath)
+    print(unsortedDicomDir)
+    print(niftiPath)
+
+    # Check for which "which" to use depending on OS
+    if os.name == 'nt':  # If OS is Windows
+        which = 'where'
+        copy = 'copy'
+    else:
+        which = 'which'
+        copy = 'cp -r'
+
+    # Make sur dcm2niix is installed
+    if os.system(which + ' dcm2niix') != 0:
+        print('Error: dcm2niix is not installed.')
+
+    # Make sure dcm2bids is installed
+    if os.system(which + ' dcm2bids') != 0:
+        raise Exception('Error: dcm2bids is not installed.')
+
+    # Create bids structure for data
+    participant = ''
+    if os.system('dcm2bids_scaffold -o ' + niftiPath) != 0:
+        raise Exception('Error: dcm2bids_scaffold')
+
+    # Add original data to niftiPath/sourcedata
+    if os.system(copy + ' ' + unsortedDicomDir + ' ' + os.path.join(niftiPath, 'sourcedata')) != 0:
+        raise Exception('Error: copy')
+
+    # Call the dcm2bids_helper
+    if os.system('dcm2bids_helper -d ' + unsortedDicomDir + ' -o ' + niftiPath) != 0:
+        raise Exception('Error: dcm2bids_helper')
+
+    # Check if there is data
+    helperPath = os.path.join(niftiPath, 'tmp_dcm2bids', 'helper')
+    if not os.path.isdir(helperPath):
+        raise Exception('Error: dcm2bids_helper could not create directory helper')
+
+    # Make sure there is data in niftiPath / tmp_dcm2bids / helper
+    helperfileList = os.listdir(helperPath)
+    if not helperfileList:
+        raise Exception('No data to process')
+
+    # Create list of acquisitions
+    acquisitionNames = list()
+    acquisitionNumbers = list()
+    iAcq = 0
+
+    # Create list containing all files
+    for iFile in range(len(helperfileList)):
+        # If it's a json file
+        name, ext = os.path.splitext(helperfileList[iFile])
+        if ext == '.json':
+            # Check for both.gz and .nii
+            booleanList = (np.in1d(helperfileList, name + '.nii') + np.in1d(helperfileList, name + '.nii.gz')).tolist()
+            niftiFile = helperfileList[booleanList.index(1)]  # Returns an exception if no True is found
+
+        # Read json file
+        _, _, jsonInfo = read_nii(os.path.join(helperPath, niftiFile))
+#     iAcq = iAcq + 1;
+#
+#     # Create future folder name acquisitionNumbers
+#     {iAcq, 1} = sprintf('#03d', jsonInfo.SeriesNumber);
+#     acquisitionNames
+#     {iAcq, 1} = jsonInfo.SeriesDescription;
+#     # ** ** ** *Modality could be acquisition name ** ** ** **
+#     modality{iAcq, 1} = jsonInfo.Modality;
```

Comment:
This loop is reading then ignoring the contents of the image files, which is very expensive. Could you expose a `read_nii_sidecard()` that doesn't touch the .nii images at all?

It also seems like you don't even need to look for the .nii files at all.

So I would redo this as

```suggestion
    
    # Create list containing all files
    for metadata in [fname for fname in os.listdir(helperPath) if fname.endswith(".json")]:
        # Read json file
        jsonInfo = read_nii_sidecar(os.path.join(helperPath, metadata))
```

But while I'm here, let me point out that python has its own vectorization built in; it's not super parallelizable like matlab or numpy but it's elegant to write:

```suggestion
    # Create list containing all files
    folder = os.listdir(helperPath)
    for file in [file for file in folder if file.endswith(".json")]:
        # Check for both.gz and .nii
        niftis = [name+ext for ext in [".nii", ".nii.gz"] if (name+ext) in folder]
        nifti = niftis[0]  # raises an exception if not found
        # or you could be more explicit:
        #if not niftis: raise ValueError(f'{file} has no associated image file')
       
        # Read json file
        _, _, jsonInfo = read_nii(os.path.join(helperPath, nifti))
```

And also by the way, you should almost never do `var+=1` in python; use a list or use a for loop. You can do `for i in itertools.count()` if you need an infinite list. It's so much more reliable to have the loop conditions set at the outside instead of partly existing inside.